# Week 8 Assessment
# Single Agent Smart Assistant using Agentic AI

## Objective

The objective of this project is to build a simple Single-Agent AI system capable of understanding user queries, routing them based on intent, invoking appropriate tools, and returning structured JSON responses. The project demonstrates the core concepts of Agentic AI, including conditional routing, tool integration, and basic error handling.

## Problem Statement

Build a Single-Agent Smart Assistant that can:

- Understand user queries
- Route tasks based on user intent
- Use external tools when required
- Return structured JSON responses

The assistant should support:

- Mathematical calculations
- Keyword extraction
- General responses

## Libraries Used

- json – for handling structured JSON responses.
- Python Built-in Functions – for implementing conditional routing and tool execution.

In [ ]:
import json

## Tool Descriptions

### Calculator Tool
This tool evaluates mathematical expressions entered by the user and returns the computed result. It also handles invalid expressions through exception handling.

### Keyword Extraction Tool
This tool extracts important keywords from a given sentence by selecting unique words with more than four characters. It helps identify the key terms in a text.

In [1]:
#TOOL 1: Calculator

def calculator(expression: str) -> str:
    """Evaluate a mathematical expression."""
    try:
        return str(eval(expression))
    except Exception:
        return "Error in calculation"

In [2]:
#TOOL 2: Keyword Extractor

def extract_keywords(text: str) -> list:
    """Extract keywords from text."""
    try:
        words = text.split()
        keywords = list(set([w.lower().strip(".,!?") for w in words if len(w) > 4]))
        return keywords[:5]
    except Exception:
        return []

## Agent Logic

The smart assistant follows a rule-based routing approach.

- If the query contains **"calculate"**, it invokes the Calculator Tool.
- If the query contains **"keywords"**, it invokes the Keyword Extraction Tool.
- Otherwise, it provides a general response.

The agent first analyzes the user query, determines the intent, invokes the appropriate tool if required, and finally returns a structured JSON response
The output is returned in a structured JSON format for consistency.

In [10]:
def agent(query: str):
    query_lower = query.lower()
    try:
        # Calculator Route
        if "calculate" in query_lower:
            expression = query_lower.replace("calculate", "").strip()
            result = calculator(expression)
            return {
                "type": "calculation",
                "result": result
            }
        # Keyword Extraction Route
        elif "keywords" in query_lower:
            text = query.replace("Extract keywords from", "").replace("extract keywords from", "").strip()
            result = extract_keywords(text)
            return {
                "type": "keywords",
                "result": result
            }
        elif "machine learning" in query_lower:
          return {
               "type": "general",
               "result": "Machine learning is a branch of Artificial Intelligence that enables computers to learn patterns from data and make predictions without being explicitly programmed."
                }
        elif "artificial intelligence" in query_lower:
          return {
                "type": "general",
                "result": "Artificial Intelligence is the simulation of human intelligence by machines to perform tasks such as learning, reasoning, and decision making."
    }
        # General Route
        else:
            return {
                "type": "general",
                "result": f"I understand your query: '{query}'. Currently I can perform calculations and keyword extraction."
            }
    except Exception as e:
        return {
            "type": "error",
            "result": str(e)
        }

## Expected Output Format

The assistant returns responses in the following JSON format:

{
    "type": "calculation / keywords / general / error",
    "result": ...
}

In [11]:
# Test Cases

queries = [
    "Calculate 20 + 5",
    "Calculate 45 * 6",
    "Extract keywords from Artificial Intelligence is transforming industries",
    "Extract keywords from Machine learning improves decision making",
    "What is machine learning?",
    "Tell me a joke",
    "Calculate 5/0"
]

for q in queries:
    print("Query:", q)
    print("Response:", agent(q))
    print("-" * 50)

Query: Calculate 20 + 5
Response: {'type': 'calculation', 'result': '25'}
--------------------------------------------------
Query: Calculate 45 * 6
Response: {'type': 'calculation', 'result': '270'}
--------------------------------------------------
Query: Extract keywords from Artificial Intelligence is transforming industries
Response: {'type': 'keywords', 'result': ['transforming', 'intelligence', 'artificial', 'industries']}
--------------------------------------------------
Query: Extract keywords from Machine learning improves decision making
Response: {'type': 'keywords', 'result': ['making', 'learning', 'machine', 'improves', 'decision']}
--------------------------------------------------
Query: What is machine learning?
Response: {'type': 'general', 'result': 'Machine learning is a branch of Artificial Intelligence that enables computers to learn patterns from data and make predictions without being explicitly programmed.'}
--------------------------------------------------
Q

In [8]:
while True:

    user_input = input("Enter query (type 'exit' to stop): ")

    if user_input.lower() == "exit":
        break

    print("Response:", agent(user_input))

Enter query (type 'exit' to stop): calculate 5-2
Response: {'type': 'calculation', 'result': '3'}
Enter query (type 'exit' to stop): What is machine learning?
Response: {'type': 'general', 'result': "I understand your query: 'What is machine learning?'. Currently I can perform calculations and keyword extraction."}
Enter query (type 'exit' to stop): Extract keywords from Machine learning improves decision making
Response: {'type': 'keywords', 'result': ['making', 'learning', 'machine', 'improves', 'decision']}
Enter query (type 'exit' to stop): exit


## Agent Workflow

```text
                User Query
                     │
                     ▼
             Intent Detection
                     │
      ┌──────────────┼──────────────┐
      ▼              ▼              ▼
 Calculation     Keyword Tool   General Query
      │              │              │
      └──────────────┼──────────────┘
                     ▼
         Structured JSON Output
```

## Experimental Results

| Query | Routed Tool | Output |
|--------|-------------|--------|
| Calculate 20 + 5 | Calculator | 25 |
| Calculate 45 × 6 | Calculator | 270 |
| Extract keywords from Artificial Intelligence is transforming industries | Keyword Tool | ['transforming', 'intelligence', 'artificial', 'industries'] |
| Extract keywords from Machine learning improves decision making | Keyword Tool | ['making', 'learning', 'machine', 'improves', 'decision'] |
| What is machine learning? | General Response | Returns a predefined explanation of Machine Learning |
| Tell me a joke | General Response | Indicates that only calculations and keyword extraction are currently supported |
| Calculate 5/0 | Calculator (Error Handling) | Error in calculation |

## Observations

- The agent successfully identified user intent using rule-based conditional routing.
- Mathematical queries were correctly processed by the Calculator Tool.
- Keyword extraction generated relevant keywords from input text.
- General knowledge queries returned predefined responses when available.
- Unsupported general queries were handled gracefully without crashing the system.
- Invalid mathematical expressions, such as division by zero, were handled using basic error handling and returned an appropriate error message.
- The agent produced structured JSON outputs for all query types, making the responses consistent and easy to interpret.

## Learning Outcomes

- Understood the concept of Agentic AI.
- Implemented rule-based intent detection.
- Integrated multiple tools within a single agent.
- Learned conditional routing techniques.
- Implemented structured JSON responses.
- Applied basic error handling for reliable execution.

## Limitations

- Uses simple keyword-based routing instead of an LLM.
- Supports only two tools.
- Does not remember previous conversations.
- Cannot handle multiple tasks in a single query.


## Future Scope

- Replace rule-based routing with an LLM-based intent classifier.
- Add memory for multi-turn conversations.
- Integrate additional tools such as Weather API, Web Search, and Translation.
- Support multiple tool calls within a single user query.
- Extend the system into a multi-agent architecture.

## Conclusion

A Single-Agent Smart Assistant was successfully developed using rule-based conditional routing and tool integration. The assistant correctly identified user intent, invoked the appropriate tool, and returned structured JSON responses. The project demonstrates the core concepts of Agentic AI, including decision-making, tool usage, and basic error handling. This implementation provides a strong foundation for developing more advanced autonomous AI agents.